Client requests hang on DB backend timeout

Hi,
We have a simple Docker test setup with 3 node PXC and ProxySQL in front of it, using mysql_galera_hostgroups for dynamic topology configuration as it should be. All settings in ProxySQL are default, nothing special.

Then we have a simple PHP application that is using mysqli standard library to run SELECT queries to ProxySQL which in turn randomly sends queries to all 3 PXC backends.

The problem arises with the following test:
We simply run `docker pause pxc-node1` which suspends mysqld process, thus simulating timeout on backend or very unresponsive DB.

At this point PHP application hangs forever. It is visible that PHP application is able to connect to ProxySQL, ProxySQL accepts connection, but then nothing happens and application waits for a response from ProxySQL forever, because real backend DB is obviously not responding with anything.

In ProxySQL logs it's clearly visible that it detected that pxc-node1 is gone with healthcheck timeout and puts it to OFFLINE_HARD.
```
2020-11-03 11:49:36 MySQL_Monitor.cpp:1549:monitor_galera_thread(): [ERROR] Error on Galera check for pxc-node1:3306 after 1002ms. Unable to create a connection. If the server is overload, increase mysql-monitor_connect_timeout. Error: timeout or error in creating new connection: Lost connection to MySQL server at 'handshake: reading inital communication packet', system error: 110.
2020-11-03 11:49:36 MySQL_Monitor.cpp:1785:monitor_galera_thread(): [ERROR] Server pxc-node1:3306 missed 3 Galera checks. Assuming offline
```

If I terminate PHP application and run it again, then it works fine and simply not sending requests to a dead node anymore as ProxySQL is not routing requests to a dead node anymore. This is expected of course.

What is not expected is that all requests and connections that were created/running during node outage are freezing the whole application without any chance to recover.

It's also worth nothing that if I run `docker unpause pxc-node1`, then application immediately unfreezes and continue going. Then ProxySQL also reports that health check restored and puts node back to ONLINE and everything looks normal again.

I tried all possible combinations of %timeout% variables in ProxySQL, but nothing helped.

IMHO, ProxySQL should send error back to client for requests that ProxySQL waiting for respond from an already dead DB, in case healthcheck marks this node as dead.

Any advices there? 

Thank you!

```
mysql_servers =
(
    { address="pxc-node1", port=3306, hostgroup=10, max_connections=100, weight=100 },
    { address="pxc-node2", port=3306, hostgroup=10, max_connections=100, weight=100  },
    { address="pxc-node3", port=3306, hostgroup=10, max_connections=100, weight=100 }
)

mysql_galera_hostgroups =
(
    {
        writer_hostgroup=10
        backup_writer_hostgroup=20
        reader_hostgroup=30
        offline_hostgroup=9999
        max_writers=1
        writer_is_also_reader=1
        max_transactions_behind=100
        active=1
    }
)


#defines MySQL Query Rules
mysql_query_rules:
(
    {
        rule_id=1
        active=1
        match_pattern="^SELECT .* FOR UPDATE$"
        destination_hostgroup=10
        apply=1
    },
    {
        rule_id=2
        active=1
        match_pattern="^SELECT"
        destination_hostgroup=30
        apply=1
    },
    {
        rule_id=3
        active=1
        match_pattern=".*"
        destination_hostgroup=10
        apply=1
    }
)
```

At 12:44:41:
```
2020-11-03 12:44:41 MySQL_Monitor.cpp:1785:monitor_galera_thread(): [ERROR] Server pxc-node1:3306 missed 3 Galera checks. Assuming offline
```

When do you run `docker unpause pxc-node1` ?

--- a/lib/mysql_connection.cpp
+++ b/lib/mysql_connection.cpp
@@ -2313,6 +2313,7 @@ int MySQL_Connection::async_send_simple_command(short event, char *stmt, unsigne
 	PROXY_TRACE();
 	assert(mysql);
 	assert(ret_mysql);
+	server_status=parent->status; // we copy it here to avoid race condition. The caller will see this
 	if (
 		(parent->status==MYSQL_SERVER_STATUS_OFFLINE_HARD) // the server is OFFLINE as specific by the user
 		||

--- a/lib/mysql_connection.cpp
+++ b/lib/mysql_connection.cpp
@@ -1616,15 +1616,9 @@ int MySQL_Connection::async_query(short event, char *stmt, unsigned long length,
 	assert(mysql);
 	assert(ret_mysql);
 	server_status=parent->status; // we copy it here to avoid race condition. The caller will see this
-	if (
-		(server_status==MYSQL_SERVER_STATUS_OFFLINE_HARD) // the server is OFFLINE as specific by the user
-		||
-		(server_status==MYSQL_SERVER_STATUS_SHUNNED && parent->shunned_automatic==true && parent->shunned_and_kill_all_connections==true) // the server is SHUNNED due to a serious issue
-		||
-		(server_status==MYSQL_SERVER_STATUS_SHUNNED_REPLICATION_LAG) // slave is lagging! see #774
-	) {
+	if (IsServerOffline())
 		return -1;
-	}
+
 	if (myds) {
 		if (myds->DSS != STATE_MARIADB_QUERY) {
 			myds->DSS = STATE_MARIADB_QUERY;
@@ -1750,6 +1744,10 @@ int MySQL_Connection::async_change_user(short event) {
 	PROXY_TRACE();
 	assert(mysql);
 	assert(ret_mysql);
+	server_status=parent->status; // we copy it here to avoid race condition. The caller will see this
+	if (IsServerOffline())
+		return -1;
+
 	switch (async_state_machine) {
 		case ASYNC_CHANGE_USER_SUCCESSFUL:
 			unknown_transaction_status = false;
@@ -1793,6 +1791,10 @@ int MySQL_Connection::async_select_db(short event) {
 	PROXY_TRACE();
 	assert(mysql);
 	assert(ret_mysql);
+	server_status=parent->status; // we copy it here to avoid race condition. The caller will see this
+	if (IsServerOffline())
+		return -1;
+
 	switch (async_state_machine) {
 		case ASYNC_INITDB_SUCCESSFUL:
 			unknown_transaction_status = false;
@@ -1830,6 +1832,10 @@ int MySQL_Connection::async_set_autocommit(short event, bool ac) {
 	PROXY_TRACE();
 	assert(mysql);
 	assert(ret_mysql);
+	server_status=parent->status; // we copy it here to avoid race condition. The caller will see this
+	if (IsServerOffline())
+		return -1;
+
 	switch (async_state_machine) {
 		case ASYNC_SET_AUTOCOMMIT_SUCCESSFUL:
 			unknown_transaction_status = false;
@@ -1869,6 +1875,10 @@ int MySQL_Connection::async_set_names(short event, unsigned int c) {
 	PROXY_TRACE();
 	assert(mysql);
 	assert(ret_mysql);
+	server_status=parent->status; // we copy it here to avoid race condition. The caller will see this
+	if (IsServerOffline())
+		return -1;
+
 	switch (async_state_machine) {
 		case ASYNC_SET_NAMES_SUCCESSFUL:
 			unknown_transaction_status = false;
@@ -1908,6 +1918,10 @@ int MySQL_Connection::async_set_option(short event, bool mask) {
 	PROXY_TRACE();
 	assert(mysql);
 	assert(ret_mysql);
+	server_status=parent->status; // we copy it here to avoid race condition. The caller will see this
+	if (IsServerOffline())
+		return -1;
+
 	switch (async_state_machine) {
 		case ASYNC_SET_OPTION_SUCCESSFUL:
 			unknown_transaction_status = false;

